astropy/io/ascii/src/tokenizer.c

```diff
@@ -872,7 +877,45 @@ double xstrtod(const char *str, char **endptr, char decimal,
     if (negative) number = -number;
 
     // Process an exponent string
-    if (toupper(*p) == toupper(sci))
+    if (autoexp)
+    {
+        // check for possible Fortran exponential notations, including
+        // triple-digits with no character
+        exp = toupper(*p);
+        if (exp == 'E' || exp == 'D' || exp == 'Q' || *p == '+' || *p == '-')
+        {
+            // Handle optional sign
+            negative = 0;
+            switch (exp)
+            {
+            case '-': negative = 1;   // Fall through to increment pos
+            case '+': p++;
+                break;
+            case 'E':
+            case 'D':
+            case 'Q':
+                switch (*++p)
+                {
+                case '-': negative = 1;   // Fall through to increment pos
+                case '+': p++;
+                }
+            }
+
+            // Process string of digits
+            n = 0;
+            while (isdigit(*p))
+            {
+                n = n * 10 + (*p - '0');
+                p++;
+            }
+
+            if (negative)
+                exponent -= n;
+            else
+                exponent += n;
+        }
+    }
+    else if (toupper(*p) == toupper(sci))
```

Comment:
Before the `if (toupper(sci) == 'A')`?
Actually one option to pull the last bit of performance I had not played with! Since in the `! autoexp` case this is only needed this single time I thought it might be faster to call both `toupper()` right there in place, (I did experiment with `exp = toupper(*p);` outside of the loop), but the timing changes seem way below measurability. If there is really need for more speed-up, one should probably rely on the caller already passing the uppercase char (as it is already doing right now).
But probably not worth the loss in fault-tolerance.